features/networking/multus.feature

```diff
@@ -55,3 +55,118 @@ Feature: Multus-CNI related scenarios
       | curl | --connect-timeout | 5 | <%= cb.pod1_multus_ip %>:8080 |
     Then the step should succeed
     And the output should contain "Hello OpenShift"
+
+
+  # @author bmeng@redhat.com
+  # @case_id OCP-21489
+  @admin
+  Scenario: Create pods with multus-cni - macvlan private mode
+    # Make sure that the multus is enabled
+    Given the master version >= "4.0"
+    And the multus is enabled on the cluster
+
+    # Create the net-attach-def via cluster admin
+    Given I have a project
+    When I run the :create admin command with:
+      | f | https://raw.githubusercontent.com/openshift-qe/v3-testfiles/master/networking/multus-cni/NetworkAttachmentDefinitions/macvlan-private.yaml |
+      | n | <%= project.name %> |
+    Then the step should succeed
+
+    # Create the first pod which consumes the macvlan custom resource
+    When I run the :create client command with:
+      | f | https://raw.githubusercontent.com/openshift-qe/v3-testfiles/master/networking/multus-cni/Pods/1interface-macvlan-private.yaml |
+    Then the step should succeed
+    And a pod becomes ready with labels:
+      | name=macvlan-private-pod |
+    And evaluation of `pod.node_name` is stored in the :pod_node clipboard
+
+    # Check that the macvlan with mode private is added to the pod
+    When I execute on the pod:
+      | /usr/sbin/ip | -d | link |
+    Then the output should contain "net1"
+    And the output should contain "macvlan mode private"
+    When I execute on the pod:
+      | bash | -c | /usr/sbin/ip -f inet addr show net1 \| grep -Po 'inet \K[\d.]+' |
+    Then the output should match "10.1.1.\d{1,3}"
+    And the expression should be true> IPAddr.new(@result[:response].chomp)
+    And evaluation of `@result[:response].chomp` is stored in the :pod1_multus_ip clipboard
+    When I execute on the pod:
+      | bash | -c | /usr/sbin/ip -f inet addr show eth0 \| grep -Po 'inet \K[\d.]+' |
+    Then the step should succeed
+    And evaluation of `@result[:response].chomp` is stored in the :pod1_sdn_ip clipboard
+
+    # Create the second pod which consumes the macvlan cr
+    When I run oc create over "https://raw.githubusercontent.com/openshift-qe/v3-testfiles/master/networking/multus-cni/Pods/1interface-macvlan-private.yaml" replacing paths:
+      | ["spec"]["nodeName"] | "<%= cb.pod_node %>" |
+    Then the step should succeed
+    And 2 pods become ready with labels:
+      | name=macvlan-private-pod |
+    And evaluation of `pod(-1).name` is stored in the :pod2 clipboard
+
+    # Try to access both the cluster ip and macvlan ip on pod1 from pod2
+    When I execute on the "<%= cb.pod2 %>" pod:
+      | curl | --connect-timeout | 5 | <%= cb.pod1_sdn_ip %>:8080 |
```

Comment:
@bmeng, what's the reason for setting the specific timeout period to 5 seconds?

Reply:
I guess he don't want curl to wait forever for response which might block the automation resources but limiting curling time to 5 seconds. I guess curl sometimes go in forever waiting state. If he don't get curl response in 5 seconds the subsequent steps will eventually fail anyhow

Reply:
is 5 seconds enough?

Reply:
Yeah, just like @anuragthehatter said, we just want it exit in a time limit if the target cannot be reached via the curl, unless it will wait for a quite long time.

And if the access succeeded, the curl request should be finished in less than 0.1 second.

Reply:
ok.  LGTM.